node/src/node_configurator/node_configurator_standard.rs

```diff
@@ -136,35 +137,152 @@ fn collect_externals_from_multi_config(
     )
 }
 
+fn extract_values_vcl_fill_multiconfig_vec(
+    config_file_vcl: Box<dyn VirtualCommandLine>,
+    environment_vcl: Box<dyn VirtualCommandLine>,
+    commandline_vcl: Box<dyn VirtualCommandLine>,
+    user_specific_data: UserSpecifiedData,
+) -> (Vec<String>, Vec<String>) {
+    let mut unspecified_vec: Vec<String> = vec!["".to_string()];
+    let mut specified_vec: Vec<String> = vec!["".to_string()];
+    let config_file_specified = user_specific_data.config_file_spec;
+    let config_file_path = user_specific_data.config_file;
+    let extract_value_from_vcl =
+        |vcl: &dyn VirtualCommandLine, name: &str, var: &str, spec: bool| {
+            let args = vcl.args();
+            let extracted_args = args
+                .iter()
+                .enumerate()
+                .filter(|(_index, vcl_arg)| vcl_arg.deref() == &name.to_string())
+                .collect_vec();
+            match extracted_args.is_empty() {
+                false => {
+                    let val_str = args[extracted_args[0].0 + 1].clone();
+                    (val_str, true)
+                }
+                true => (var.to_string(), spec),
+            }
+        };
+    let (cf_real_user, cf_real_user_specified) = extract_value_from_vcl(
+        &*config_file_vcl,
+        "--real-user",
+        user_specific_data.real_user.to_string().as_str(),
+        user_specific_data.real_user_spec,
+    );
+
+    let (env_real_user, env_real_user_specified) = extract_value_from_vcl(
+        &*environment_vcl,
+        "--real-user",
+        user_specific_data.real_user.to_string().as_str(),
+        user_specific_data.real_user_spec,
+    );
+
+    let (cmd_real_user, cmd_real_user_specified) = extract_value_from_vcl(
+        &*commandline_vcl,
+        "--real-user",
+        user_specific_data.real_user.to_string().as_str(),
+        user_specific_data.real_user_spec,
+    );
+
+    let mut fill_specified_or_unspecified_box =
+        |key: &str, value: &str, specified: bool| match value.is_empty() {
+            true => (),
+            false => match specified {
+                true => match specified_vec.contains(&key.to_string()) {
```

Comment:
I see some code duplication here. You already have some local functions; you can probably pull out the `match [un]specified_vec` part into another one.

Also, the do-this-if-it-already-exists-or-this-if-it-doesn't logic is already part of `HashMap`, if you decided to use that. Unfortunately, what you need to make a VCL is a vector. However, there might be an easy way to convert a `HashMap` into a `Vec` with pairs. I don't know, but it might be worth looking into to simplify this code.

Reply:
Turns out I think you can do it by calling `.iter()` on the `HashMap` and then `fold()` on the result, starting with an empty vector. Every time through the closure you pass to `fold()`, push the two items in the pair onto the vector; that should do it, maybe in two or three lines.